### PR TITLE
[Shapes] OCaml 4.14 + Shapes

### PR DESCRIPTION
**Pull request by voodoos**
Another Pr for early testing.
- Based on #1415, an up-to-date upgrade to OCaml 4.14
- With the latest version of @gasche's [strong call-by-need PR](https://github.com/ocaml/ocaml/pull/10825)
- With the locate code that uses shapes similar to others recent testing PRs

- [x] Most tests are improved, one is not (jumping to declaration instead of a functor argument).
- [x] Documentation tests are not passing at all yet.
- [x] Add tests illustrating module aliases issue
- ~[ ] fix by patching the compiler~
- [x] fix documentation and dot-pp-dot-ml-dune tests 

The corresponding compiler (4.14 + @gasche's cbn) to be used with this merlin can be found here:
https://github.com/voodoos/ocaml/tree/shape-strong-call-by-need

### CODE REVIEW

**Comment by voodoos** [comment]
@trefis all tests are now passing, and jumping to the argument of a functor works again. This PR is ready for review !

To restore that behaviour I had to store the functor arguments uid in the `uid_to_loc` table of the local environment. See c5e6d07.
I suppose it is not necessary to upstream this as it only occurs in local buffers. But we might want to to reduce the divergence between Merlin and the compiler... What do you think ?

**Comment by voodoos** [comment]
@trefis : this now use uids to answer document queries when the target is in another comp_unit. It fallsback on the old heuristic when the target is in the current buffer or the previous algorithm found nothing (necessary for test `unattached-comment`).

I also reverted the changes for functor arguments. We now jump to the correct place in the argument's signature.

**Comment by voodoos** [comment]
I rebased both the 414 branch and this PR.
This led to the discovery of an issue with module aliases.
I will add more tests to illustrate the issue.

- [ ] Add tests illustrating module aliases issue
- [ ] fix by patching the compiler
- [ ] fix documentation and dot-pp-dot-ml-dune tests

**Comment by voodoos** [comment]
@trefis all test fixed. This is ready for review. My original intent was to merge this into the 4.14 update PR but maybe we can first merge the update PR onto master and this one second ?

- `get_doc` now use uids to get docstrings from external compilation units. I took that opportunity to fix https://github.com/ocaml/merlin/issues/1421.

File: tests/test-dirs/locate/locate-stdlib.t [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  > EOF
+
+  $ cat >dune <<EOF
+  > (executable (name s))
+  > EOF
+
+  $ dune build ./s.exe
+
+Should jump to set.ml:
+  $ $MERLIN single locate -look-for ml -position 5:13 -filename ./s.ml < ./s.ml
```
Comment (trefis) [inline tests/test-dirs/locate/locate-stdlib.t:20]: That's the kind of test whose output is likely going to be unstable: `set.ml` might change with each new release of ocaml, moving a bit the position returned by locate, creating noise in the diff we'll have to check.

What are you trying to test here and can we test it without adding "external dependencies"?

File: tests/test-dirs/locate/partial-cmt.t/run.t [hunk 2]
```diff
@@ -48,9 +48,10 @@ That is: if the file is a.mli then the test is broken:
...
       "pos": {
         "line": 1,
-        "col": 0
+        "col": 9
       }
     },
     "notifications": []
   }
 
+TODO With shape it is extected that this test will fail.
```
Comment (trefis) [inline tests/test-dirs/locate/partial-cmt.t/run.t:57]: expected*.

Also: let's just remove the test if we do not expect things to work with partial cmts.

Comment (voodoos) [reply tests/test-dirs/locate/partial-cmt.t/run.t:57]: I kept it as an incentive :-)

**Pull request merged**